### ISSUE TITLE
Add support for CollectionInterface::reduce()

### ISSUE THREAD

**Issue by ramsey** [ramsey/collection#79]
## Background/problem

The library supports `map()` on CollectionInterface, but there is no `reduce()`.

## Proposal/solution

Add `CollectionInterface::reduce()` with the following signature:

```php
/**
 * @template TCarry
 *
 * @param callable(TCarry, T):TCarry $callback A callable to apply to each
 *     item of the collection to reduce it to a single value.
 * @param TCarry|null $initial If provided, this is the initial value provided
 *     to the callback.
 *
 * @return TCarry
 */
public function reduce(callable $callback, mixed $initial = null): mixed;
```

## Additional context

Since this is a change to the interface, it requires a major version bump.

**Comment by drupol** [comment]
Find an interface (_among some others_) for this here: https://github.com/loophp/collection/blob/master/src/Contract/Operation/FoldLeftable.php

**Comment by ramsey** [comment]
Does "fold left" mean the same as "reduce?" I'm not familiar with the phrase "fold left."

**Comment by drupol** [comment]
Yes, see https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29

**Comment by ramsey** [comment]
Implemented in 0ca6ba5f4c3bc4f29e41b39c471a4e054155a5d7
